types/lib/stdlib.go

```diff
@@ -0,0 +1,269 @@
+}
+
+type listMapper struct {
+	key    sdk.StoreKey
+	cdc    *wire.Codec
+	prefix string
+	lk     []byte
+}
+
+func NewListMapper(cdc *wire.Codec, key sdk.StoreKey, prefix string) ListMapper {
+	lk, err := cdc.MarshalBinary(int64(-1))
+	if err != nil {
+		panic(err)
+	}
+	return listMapper{
+		key:    key,
+		cdc:    cdc,
+		prefix: prefix,
+		lk:     lk,
+	}
+}
+
+func (lm listMapper) Len(ctx sdk.Context) int64 {
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.LengthKey())
+	if bz == nil {
+		zero, err := lm.cdc.MarshalBinary(0)
+		if err != nil {
+			panic(err)
+		}
+		store.Set(lm.LengthKey(), zero)
+		return 0
+	}
+	var res int64
+	if err := lm.cdc.UnmarshalBinary(bz, &res); err != nil {
+		panic(err)
+	}
+	return res
+}
+
+func (lm listMapper) Get(ctx sdk.Context, index int64, ptr interface{}) {
+	if index < 0 {
+		panic(fmt.Errorf("Invalid index in ListMapper.Get(ctx, %d, ptr)", index))
+	}
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.ElemKey(index))
+	if err := lm.cdc.UnmarshalBinary(bz, ptr); err != nil {
+		panic(err)
+	}
+}
+
+func (lm listMapper) Set(ctx sdk.Context, index int64, value interface{}) {
+	if index < 0 {
+		panic(fmt.Errorf("Invalid index in ListMapper.Set(ctx, %d, value)", index))
+	}
+	store := ctx.KVStore(lm.key)
+	bz, err := lm.cdc.MarshalBinary(value)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(lm.ElemKey(index), bz)
+}
+
+func (lm listMapper) Delete(ctx sdk.Context, index int64) {
+	if index < 0 {
+		panic(fmt.Errorf("Invalid index in ListMapper.Delete(ctx, %d)", index))
+	}
+	store := ctx.KVStore(lm.key)
+	store.Delete(lm.ElemKey(index))
+}
+
+func (lm listMapper) Push(ctx sdk.Context, value interface{}) {
+	length := lm.Len(ctx)
+	lm.Set(ctx, length, value)
+
+	store := ctx.KVStore(lm.key)
+	store.Set(lm.LengthKey(), marshalInt64(lm.cdc, length+1))
+}
+
+func (lm listMapper) Iterate(ctx sdk.Context, ptr interface{}, fn func(sdk.Context, int64) bool) {
+	length := lm.Len(ctx)
+	for i := int64(0); i < length; i++ {
+		lm.Get(ctx, i, ptr)
+		if fn(ctx, i) {
+			break
+		}
+	}
+}
+
+func (lm listMapper) LengthKey() []byte {
+	return []byte(fmt.Sprintf("%s/%d", lm.prefix, lm.lk))
+}
+
+func (lm listMapper) ElemKey(i int64) []byte {
+	return []byte(fmt.Sprintf("%s/%d", lm.prefix, i))
+}
+
+// QueueMapper is a Mapper interface that provides queue-like functions
+// It panics when the element type cannot be (un/)marshalled by the codec
+
+type QueueMapper interface {
+	Push(sdk.Context, interface{})
+	// Popping/Peeking on an empty queue will cause panic
+	// The user should check IsEmpty() before doing any actions
+	Peek(sdk.Context, interface{})
+	Pop(sdk.Context)
+	IsEmpty(sdk.Context) bool
+	// Iterate() removes elements it processed; return true in the continuation to break
+	Iterate(sdk.Context, interface{}, func(sdk.Context) bool)
+	Info(sdk.Context) QueueInfo
+}
+
+type queueMapper struct {
+	key    sdk.StoreKey
+	cdc    *wire.Codec
+	prefix string
+	lm     ListMapper
+	lk     []byte
```

Comment:
Why are `lk` and `ik` struct fields instead of constants? They're set to constants in `NewQueueMapper`.